plutus-script-utils/src/Plutus/V1/Utils/Scripts.hs

```diff
@@ -0,0 +1,93 @@
+module Plutus.V1.Utils.Scripts
+    ( PV1.Datum
+    , PV1.DatumHash
+    , PV1.Redeemer
+    , PV1.RedeemerHash
+    , PV1.Validator
+    , PV1.ValidatorHash
+    , PV1.MintingPolicy
+    , PV1.MintingPolicyHash
+    , PV1.StakeValidator
+    , PV1.StakeValidatorHash
+    , datumHash
+    , redeemerHash
+    , validatorHash
+    , mintingPolicyHash
+    , stakeValidatorHash
+    , toCardanoApiScript
+    , scriptHash
+    , dataHash
+    , toCardanoAPIData
+    ) where
+
+import Cardano.Api qualified as Script
+import Cardano.Api.Shelley qualified as Script
+import Codec.Serialise (serialise)
+import Data.ByteString.Lazy qualified as BSL
+import Data.ByteString.Short qualified as SBS
+import Plutus.V1.Ledger.Scripts qualified as PV1
+import PlutusTx.Builtins qualified as Builtins
+
+-- | Hash a 'PV1.Datum builtin data.
+datumHash :: PV1.Datum -> PV1.DatumHash
+datumHash = PV1.DatumHash . dataHash . PV1.getDatum
+
+-- | Hash a 'PV1.Redeemer' builtin data.
+redeemerHash :: PV1.Redeemer -> PV1.RedeemerHash
+redeemerHash = PV1.RedeemerHash . dataHash . PV1.getRedeemer
+
+-- | Hash a 'PV1.Validator' script.
+validatorHash :: PV1.Validator -> PV1.ValidatorHash
+validatorHash =
+    PV1.ValidatorHash
+  . PV1.getScriptHash
+  . scriptHash
+  . PV1.getValidator
+
+-- | Hash a 'PV1.MintingPolicy' script.
+mintingPolicyHash :: PV1.MintingPolicy -> PV1.MintingPolicyHash
+mintingPolicyHash =
+    PV1.MintingPolicyHash
+  . PV1.getScriptHash
+  . scriptHash
+  . PV1.getMintingPolicy
+
+-- | Hash a 'PV1.StakeValidator' script.
+stakeValidatorHash :: PV1.StakeValidator -> PV1.StakeValidatorHash
+stakeValidatorHash =
+    PV1.StakeValidatorHash
+  . PV1.getScriptHash
+  . scriptHash
+  . PV1.getStakeValidator
+
+-- | Hash a 'Builtins.BuiltinData'
+dataHash :: Builtins.BuiltinData -> Builtins.BuiltinByteString
+dataHash =
+    Builtins.toBuiltin
+    . Script.serialiseToRawBytes
+    . Script.hashScriptData
+    . toCardanoAPIData
+
+-- | Convert a 'Builtins.BuiltinsData' value to a 'cardano-api' script
+--   data value.
+toCardanoAPIData :: Builtins.BuiltinData -> Script.ScriptData
+toCardanoAPIData = Script.fromPlutusData . Builtins.builtinDataToData
+
+-- | Hash a 'Script'
+scriptHash :: PV1.Script -> PV1.ScriptHash
+scriptHash =
+    PV1.ScriptHash
+    . Builtins.toBuiltin
+    . Script.serialiseToRawBytes
+    . Script.hashScript
+    . toCardanoApiScript
+
+-- | Convert a 'Script' to a 'cardano-api' script
+toCardanoApiScript :: PV1.Script -> Script.Script Script.PlutusScriptV1
```

Comment:
That's right, we should not depend on it. However, there's 2 options to compute the hash of a script:

- Depend on `cardano-ledger` instead and use their `hashScript` function
- Copy the hashScript function in `plutus-script-utils` and depend on `cardano-crypto-class`

Not sure which is the more reasonable. What do you think?

Reply:
That's a pretty good argument. And tbh I guess anyone who is *actually* using scripts *is* going to depend on `cardano-api`, so maybe it's fine.

(This is also a good argument for keeping this package in `plutus-apps` rather than `plutus`!)

Reply:
What do you think @ch1bo?

Reply:
We could live with all three options.
* `cardano-crypto-class` is very low-level, but many packages depend on it already
* `cardano-ledger` is an odd one, we only depend on it because we use it to validate transactions
* `cardano-api` is now our go-to package for transaction construction, and likely will be used by anyone embedding plutus scripts

With regards to the functions in this module:
- `toCardanoApiScript` feels odd in `plutus-script-utils` and I would expect rather a `fromPlutusScript` in `cardano-api` (we polyfilled one in our `hydra-cardano-api`)
- We were quite surprised that `plutus` (and now `plutus-apps`) does not itself define how a `Validator` is turned into a `ValidatorHash`? I would expect that to even live in `plutus` and `cardano-ledger` use `plutus`' implementation to hash scripts instead of the other way around?

Reply:
> We were quite surprised that plutus (and now plutus-apps) does not itself define how a Validator is turned into a ValidatorHash?

That's just the truth! It's not a `plutus` responsibility: we define the serialization, but it's up to the ledger/node to decide on things like what the hash function is.

> toCardanoApiScript feels odd in plutus-script-utils and I would expect rather a fromPlutusScript in cardano-api

`cardano-api` doesn't know about our script types etc. It just knows about serialized scripts as bytestrings...